### DISCUSSION TITLE
fcitx5 能否和基于wlroots 0.18 的wayland compositor 一起工作？

### DISCUSSION THREAD

**Discussion by guyuming76** [fcitx/fcitx5#1222] [Q&A] [answered]
我今天把中文输入集成加进了 基于 wlroots 0.18 的 dwl 0.7， https://gitee.com/guyuming76/dwl/tree/0.7cn/ 

启动 fcitx 前，dwl 0.7 是可以用的，但启动 fcitx5 后就不行了， dwl log 显示 keypress 事件代码被不停地调用，但我此时并没有按键盘. 我不确定是我代码的问题还是 fcitx5 和 wlroots 0.18 不兼容。

我的改动在 dwl 0.5 上一直是可以用的， dwl 0.5 到 0.7 改了多少我没仔细看，只注意到之前用 wlr_keyboard 的地方现在改用 wlr_keyboard_group 了，我的 IM 相关代码里也作了相应调整。

fcitx5 会有类似下面日志不停地输出：
[ 701183.138] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576408, 64, 0, 0, 0)
[ 701183.175] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701183.201] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576409, 4, 0, 0, 0)
[ 701183.236] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(4, 0, 0, 0)
[ 701183.257] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576410, 64, 0, 0, 0)
[ 701183.288] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701183.310] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576411, 209675241, 28, 0)
D2025-01-01 02:33:19.940973 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.941057 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.941278 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701183.712] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209675241, 28, 0)
[ 701183.737] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576412, 209665820, 28, 0)
D2025-01-01 02:33:19.941400 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.941482 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.941700 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701184.134] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209665820, 28, 0)
[ 701184.159] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576413, 8, 0, 0, 0)
[ 701184.192] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
[ 701184.213] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576414, 0, 0, 0, 0)
[ 701184.244] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701184.265] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576415, 12, 0, 0, 0)
[ 701184.295] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(12, 0, 0, 0)
[ 701184.317] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576416, 8, 0, 0, 0)
[ 701184.347] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
D2025-01-01 02:33:19.942033 waylandeventreader.cpp:24] wl_display_flush
[ 701188.914] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576417, 0, 0, 0, 0)
[ 701188.957] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701188.984] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576418, 209720502, 60, 0)
D2025-01-01 02:33:19.946657 waylandimserverv2.cpp:468] F2 IsRelease=1
D2025-01-01 02:33:19.946787 instance.cpp:919] KeyEvent: Key(F2 states=0) rawKey: Key(F2 states=0) origKey: Key(F2 states=0) Release:1 keycode: 68
D2025-01-01 02:33:19.947126 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701189.601] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209720502, 60, 0)
[ 701189.631] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576419, 64, 0, 0, 0)
[ 701189.669] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701189.695] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576420, 4, 0, 0, 0)
[ 701189.730] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(4, 0, 0, 0)
[ 701189.756] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576421, 64, 0, 0, 0)
[ 701189.791] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701189.816] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576422, 209675241, 28, 0)
D2025-01-01 02:33:19.947488 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.947609 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.947940 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701190.414] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209675241, 28, 0)
[ 701190.444] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576423, 209665820, 28, 0)
D2025-01-01 02:33:19.948116 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.948238 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.948567 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701191.041] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209665820, 28, 0)
[ 701191.070] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576424, 8, 0, 0, 0)
[ 701191.107] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
[ 701191.133] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576425, 0, 0, 0, 0)
[ 701191.167] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701191.192] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576426, 12, 0, 0, 0)
[ 701191.227] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(12, 0, 0, 0)
[ 701191.252] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576427, 8, 0, 0, 0)
[ 701191.286] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
D2025-01-01 02:33:19.948978 waylandeventreader.cpp:24] wl_display_flush
[ 701192.219] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576428, 0, 0, 0, 0)
[ 701192.259] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701192.283] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576429, 209720502, 60, 0)
D2025-01-01 02:33:19.949950 waylandimserverv2.cpp:468] F2 IsRelease=1
D2025-01-01 02:33:19.950037 instance.cpp:919] KeyEvent: Key(F2 states=0) rawKey: Key(F2 states=0) origKey: Key(F2 states=0) Release:1 keycode: 68
D2025-01-01 02:33:19.950368 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701192.841] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209720502, 60, 0)
[ 701212.858] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576430, 64, 0, 0, 0)
[ 701212.925] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701212.948] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576431, 4, 0, 0, 0)
[ 701212.979] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(4, 0, 0, 0)
[ 701213.000] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576432, 64, 0, 0, 0)
[ 701213.030] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701213.052] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576433, 209675241, 28, 0)
D2025-01-01 02:33:19.970731 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.970871 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.971143 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701213.581] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209675241, 28, 0)
[ 701213.607] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576434, 209665820, 28, 0)
D2025-01-01 02:33:19.971270 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.971353 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.971579 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701214.049] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209665820, 28, 0)
[ 701214.079] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576435, 8, 0, 0, 0)
[ 701214.116] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
[ 701214.142] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576436, 0, 0, 0, 0)
[ 701214.177] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701214.203] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576437, 12, 0, 0, 0)
[ 701214.237] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(12, 0, 0, 0)
[ 701214.263] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576438, 8, 0, 0, 0)
[ 701214.297] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
D2025-01-01 02:33:19.972009 waylandeventreader.cpp:24] wl_display_flush
[ 701216.779] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576439, 0, 0, 0, 0)
[ 701216.837] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701216.860] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576440, 209720502, 60, 0)
D2025-01-01 02:33:19.974529 waylandimserverv2.cpp:468] F2 IsRelease=1
D2025-01-01 02:33:19.974632 instance.cpp:919] KeyEvent: Key(F2 states=0) rawKey: Key(F2 states=0) origKey: Key(F2 states=0) Release:1 keycode: 68
D2025-01-01 02:33:19.974870 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701217.306] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209720502, 60, 0)
[ 701217.334] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576441, 64, 0, 0, 0)
[ 701217.367] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701217.389] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576442, 4, 0, 0, 0)
[ 701217.420] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(4, 0, 0, 0)
[ 701217.442] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576443, 64, 0, 0, 0)
[ 701217.472] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701217.494] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576444, 209675241, 28, 0)
D2025-01-01 02:33:19.975157 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.975241 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.975461 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701217.896] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209675241, 28, 0)
[ 701217.922] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576445, 209665820, 28, 0)
D2025-01-01 02:33:19.975585 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.975667 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.975885 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701218.318] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209665820, 28, 0)
[ 701218.344] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576446, 8, 0, 0, 0)
[ 701218.377] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
[ 701218.398] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576447, 0, 0, 0, 0)
[ 701218.429] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701218.450] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576448, 12, 0, 0, 0)
[ 701218.481] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(12, 0, 0, 0)
[ 701218.502] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576449, 8, 0, 0, 0)
[ 701218.533] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
D2025-01-01 02:33:19.976222 waylandeventreader.cpp:24] wl_display_flush
[ 701222.248] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576450, 0, 0, 0, 0)
[ 701222.292] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701222.319] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576451, 209720502, 60, 0)
D2025-01-01 02:33:19.979993 waylandimserverv2.cpp:468] F2 IsRelease=1
D2025-01-01 02:33:19.980117 instance.cpp:919] KeyEvent: Key(F2 states=0) rawKey: Key(F2 states=0) origKey: Key(F2 states=0) Release:1 keycode: 68
D2025-01-01 02:33:19.980451 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701222.922] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209720502, 60, 0)
[ 701222.953] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576452, 64, 0, 0, 0)
[ 701222.992] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701223.018] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576453, 4, 0, 0, 0)
[ 701223.054] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(4, 0, 0, 0)
[ 701223.080] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576454, 64, 0, 0, 0)
[ 701223.115] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701223.141] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576455, 209675241, 28, 0)
D2025-01-01 02:33:19.980812 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.980930 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.981240 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701223.674] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209675241, 28, 0)
[ 701223.700] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576456, 209665820, 28, 0)
D2025-01-01 02:33:19.981364 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.981447 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.981664 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701224.098] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209665820, 28, 0)
[ 701224.123] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576457, 8, 0, 0, 0)
[ 701224.156] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
[ 701224.178] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576458, 0, 0, 0, 0)
[ 701224.208] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701224.228] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576459, 12, 0, 0, 0)
[ 701224.259] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(12, 0, 0, 0)
[ 701224.280] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576460, 8, 0, 0, 0)
[ 701224.311] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)
D2025-01-01 02:33:19.982002 waylandeventreader.cpp:24] wl_display_flush
[ 701228.915] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576461, 0, 0, 0, 0)
[ 701228.960] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(0, 0, 0, 0)
[ 701228.986] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576462, 209720502, 60, 0)
D2025-01-01 02:33:19.986661 waylandimserverv2.cpp:468] F2 IsRelease=1
D2025-01-01 02:33:19.986792 instance.cpp:919] KeyEvent: Key(F2 states=0) rawKey: Key(F2 states=0) origKey: Key(F2 states=0) Release:1 keycode: 68
D2025-01-01 02:33:19.987123 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701229.594] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209720502, 60, 0)
[ 701229.625] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576463, 64, 0, 0, 0)
[ 701229.662] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701229.688] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576464, 4, 0, 0, 0)
[ 701229.724] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(4, 0, 0, 0)
[ 701229.749] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576465, 64, 0, 0, 0)
[ 701229.785] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(64, 0, 0, 0)
[ 701229.810] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576466, 209675241, 28, 0)
D2025-01-01 02:33:19.987481 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.987601 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.987925 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701230.394] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209675241, 28, 0)
[ 701230.424] {Default Queue} zwp_input_method_keyboard_grab_v2#20.key(576467, 209665820, 28, 0)
D2025-01-01 02:33:19.988095 waylandimserverv2.cpp:468] Super+Return IsRelease=1
D2025-01-01 02:33:19.988213 instance.cpp:919] KeyEvent: Key(Super+Return states=64) rawKey: Key(Super+Return states=64) origKey: Key(Super+Return states=64) Release:1 keycode: 36
D2025-01-01 02:33:19.988535 inputcontext.cpp:321] KeyEvent handling time: 0ms result:0
[ 701231.009] {Default Queue}  -> zwp_virtual_keyboard_v1#15.key(209665820, 28, 0)
[ 701231.038] {Default Queue} zwp_input_method_keyboard_grab_v2#20.modifiers(576468, 8, 0, 0, 0)
[ 701231.074] {Default Queue}  -> zwp_virtual_keyboard_v1#15.modifiers(8, 0, 0, 0)

**Comment by wengxt** [comment, marked as answer]
有什么问题自己检查

这个看起来像是输入法发送过去的按键（virtual_keyboard.key）你又再次发回来给输入法（创建的keygrab）死循环了

又或者是你发给fcitx的按键触发了按键重复（需要收到key release来停止）但是没有

**Reply by guyuming76** [reply]
多谢支持！原来dwl0.7里面为virtualkeyboard建立了单独的keyboard group, 这样，原来识别virtualkeyboard以便过滤事件的代码失效了。 我调整了一下，凡是不是全局变量 kb_group, 也就是凡是virtualkeyboard的事件都不转发，貌似可以工作了：   https://gitee.com/guyuming76/dwl/commit/8f68de9e7d99d93bfd21c2b29c4b9125b4d0a13b
